README.md
# GethPrograms

Contains two programs about Go-Ethereum:

1.Transaction monitoring, sending and receiving.

2.Recharge and withdrawal platform.

Work of summer internship, 2020


chain_10/history
eth.accounts
eth.pendingTransactions
miner.setEtherbase(eth.account[0])
miner.setEtherbase(eth.accounts[0])
miner.start(2);admin.sleepBlocks(10);miner.stop()
miner.setEtherbase(eth.accounts[0])
eth.pendingTransactions
eth.accounts

myEthereum/main.go
package main

import (
	"fmt"

	"github.com/ethereum/go-ethereum/ethclient"
)

var curuser string
var keystoremap map[string]string

func main() {
	keystoremap = make(map[string]string)
	keystoremap["0xc0093215bec3cbb9522352dcb4e3fa8fd5b665d1"] = `D:\Projects\GethProgram\chain_10\keystore\UTC--2020-07-30T12-08-00.423453600Z--c0093215bec3cbb9522352dcb4e3fa8fd5b665d1`
	keystoremap["0x8bf8fb9cfd048de2645539f5007d12ae6465623b"] = `D:\Projects\GethProgram\chain_10\keystore\UTC--2020-07-31T01-19-22.030531400Z--8bf8fb9cfd048de2645539f5007d12ae6465623b`
	keystoremap["0x297fca8450553a3c9103dd410b59e79cdf6f4a0a"] = `D:\Projects\GethProgram\chain_10\keystore\UTC--2020-07-31T01-19-25.477316000Z--297fca8450553a3c9103dd410b59e79cdf6f4a0a`
	keystoremap["0x85bdd4c54fc8689e2973ab8ded1cf46c55243e05"] = `D:\Projects\GethProgram\chain_10\keystore\UTC--2020-07-31T01-19-28.576035000Z--85bdd4c54fc8689e2973ab8ded1cf46c55243e05`
	keystoremap["0x7e5befa8b17710f881f9011425a93e47dfa8590e"] = `D:\Projects\GethProgram\chain_10\keystore\UTC--2020-07-31T01-19-32.013843100Z--7e5befa8b17710f881f9011425a93e47dfa8590e`
	keystoremap["0xf81233b607aa4178ed9e9e1e1a876b683619ef47"] = `D:\Projects\GethProgram\chain_10\keystore\UTC--2020-07-31T01-19-35.118545500Z--f81233b607aa4178ed9e9e1e1a876b683619ef47`

	var option int
	islogin := false
	isadmin := false

	for !islogin {
		fmt.Println("please choose your option:")
		fmt.Println("0: log in\t1: register\t2: log in as admin\t3: exit")

		_, err := fmt.Scanln(&option)
		if err != nil {
			fmt.Println("invalid input")
			continue
		}

		switch option {
		case 0:
			islogin = Login()
			fmt.Println("")
		case 1:
			Register()
			fmt.Println("")
		case 2:
			if LoginAsAdmin() {
				islogin = true
				isadmin = true
			}
			fmt.Println("")
		case 3:
			fmt.Println("")
			return
		default:
			fmt.Println("invalid input")
			fmt.Println("")
		}
	}

	client, err := ethclient.Dial(RPCAddress)
	if err != nil {
		fmt.Println("connect failededed: ", err)
		return
	}

	for !isadmin {
		fmt.Println("please choose your option:")
		fmt.Println("0: check balance\t1: recharge\t2: withdraw\t3: exit")

		_, err := fmt.Scanln(&option)
		if err != nil {
			fmt.Println("invalid input")
			continue
		}

		switch option {
		case 0:
			CheckBalance(client)
			fmt.Println("")
		case 1:
			Recharge(client)
			fmt.Println("")
		case 2:
			Withdraw(client)
			fmt.Println("")
		case 3:
			fmt.Println("")
			return
		default:
			fmt.Println("invalid input")
			fmt.Println("")
		}
	}

	for isadmin {
		fmt.Println("please choose your option:")
		fmt.Println("0: centralize\t1: exit")

		_, err := fmt.Scanln(&option)
		if err != nil {
			fmt.Println("invalid input")
			continue
		}

		switch option {
		case 0:
			Centralize(client)
			fmt.Println("")
		case 1:
			fmt.Println("")
			return
		default:
			fmt.Println("invalid input")
			fmt.Println("")
		}
	}
}


myEthereum/user.go
package main

import (
	"bufio"
	"crypto/ecdsa"
	"fmt"
	"math/big"
	"os"
	"strings"

	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/ethclient"
)

// Login : user log in
func Login() bool {
	fmt.Println("please input your mnemonic:")
	reader := bufio.NewReader(os.Stdin)
	mnemonic, _, _ := reader.ReadLine()

	if LoginAnAccount(string(mnemonic)) {
		fmt.Println("welcome, " + curuser)
		return true
	}

	fmt.Println("failed to log in: wrong mnemonic")
	return false
}

// LoginAsAdmin : log in as an admin
func LoginAsAdmin() bool {
	var password string
	fmt.Println("please input admin password:")
	fmt.Scanln(&password)

	if password != "admin" {
		fmt.Println("wrong password")
		return false
	}

	return true
}

// Register : user registers
func Register() {
	_, mnemonic, err := RegisterAnAccount()
	if err != nil {
		fmt.Println(err)
		return
	}

	fmt.Printf("succeeded to register, your mnemonic is:\n%s\n", *mnemonic)
}

// CheckBalance : check the balance in current account
func CheckBalance(client *ethclient.Client) {
	_, balance, err := RefreshAccountInfo(client, curuser)
	if err != nil {
		fmt.Println(err)
		return
	}

	fmt.Printf("your balance is:\n%s\n", *balance)
}

// Recharge : recharge
func Recharge(client *ethclient.Client) {
	// get privateKey
	var privateKeyFile, password string
	fmt.Println("please input your key file path:")
	fmt.Scanln(&privateKeyFile)
	fmt.Println("please input your password:")
	fmt.Scanln(&password)

	keyValue, err := GetPrivateKey(&privateKeyFile, &password)
	if err != nil {
		fmt.Println("failed to get privateKey: ", err)
		return
	}

	privateKey, err := crypto.HexToECDSA(*keyValue)
	if err != nil {
		fmt.Println("failed to get privateKey: ", err)
		return
	}

	publicKey := privateKey.Public()
	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
	if !ok {
		fmt.Println("Error casting public key to ECDSA.")
		return
	}

	fromAddress := crypto.PubkeyToAddress(*publicKeyECDSA)
	ethaddress := fromAddress.Hex()

	// get value
	var valuestr string
	fmt.Println("please input the value:")
	fmt.Scanln(&valuestr)

	value := new(big.Int)
	value, ok = value.SetString(valuestr, 10)
	if !ok {
		fmt.Println("invalid input")
		return
	}

	// send transaction
	hash, err := StartATransaction(client, value, ethaddress, curuser, privateKey)
	if err != nil {
		fmt.Println("fail to recharge: ", err)
		return
	}

	// save transaction
	err = SaveATransaction(curuser, *hash, "0", valuestr)
	if err != nil {
		fmt.Println("fail to save transaction: ", err)
		return
	}

	fmt.Printf("transaction submitted:%v\n", *hash)
}

// Withdraw : Withdraw
func Withdraw(client *ethclient.Client) {
	// get privateKey
	mk := keystoremap[strings.ToLower(MainAddress)]
	mp := "admin"
	keyValue, err := GetPrivateKey(&mk, &mp)
	if err != nil {
		fmt.Println("failed to get privateKey: ", err)
		return
	}

	privateKey, err := crypto.HexToECDSA(*keyValue)
	if err != nil {
		fmt.Println("failed to get privateKey: ", err)
		return
	}

	// get address
	var ethaddress string
	fmt.Println("please input your ethereum address:")
	fmt.Scanln(&ethaddress)

	// get value
	var valuestr string
	fmt.Println("please input the value:")
	fmt.Scanln(&valuestr)

	value := new(big.Int)
	value, ok := value.SetString(valuestr, 10)
	if !ok {
		fmt.Println("invalid input")
		return
	}

	// get balance
	_, balancestr, err := RefreshAccountInfo(client, curuser)
	if err != nil {
		fmt.Println(err)
		return
	}

	balance := new(big.Int)
	balance, ok = balance.SetString(*balancestr, 10)
	if !ok {
		fmt.Println("invalid input")
		return
	}

	// compare balance to value
	if balance.Cmp(value) == -1 {
		fmt.Println("balance is not enough")
		return
	}

	// send transaction
	hash, err := StartATransaction(client, value, MainAddress, ethaddress, privateKey)
	if err != nil {
		fmt.Println("fail to recharge: ", err)
		return
	}

	// save transaction
	err = SaveATransaction(curuser, *hash, "1", valuestr)
	if err != nil {
		fmt.Println("fail to save transaction: ", err)
		return
	}

	fmt.Printf("transaction submitted: %v\n", *hash)
}

// Centralize : centralize all the balance in user accounts
func Centralize(client *ethclient.Client) {
	// refresh all accounts
	info, ok := RefreshAllAccount(client)
	if !ok {
		fmt.Println("failed to refresh balance")
		return
	}

	fmt.Println("refresh accounts done")

	// send transactions
	for _, userinfo := range *info {
		// if address balance is 0, then continue
		if userinfo["addressbalance"] == "0" {
			continue
		}

		// get privateKey
		mk := keystoremap[strings.ToLower(userinfo["address"])]
		mp := "admin"

		keyValue, err := GetPrivateKey(&mk, &mp)
		if err != nil {
			fmt.Println("failed to get privateKey: ", err)
			return
		}

		privateKey, err := crypto.HexToECDSA(*keyValue)
		if err != nil {
			fmt.Println("failed to get privateKey: ", err)
			return
		}

		// get value
		value := new(big.Int)
		value, ok := value.SetString(userinfo["addressbalance"], 10)
		if !ok {
			fmt.Println("invalid input")
			return
		}

		// send transaction
		hash, err := StartATransaction(client, value, userinfo["address"], MainAddress, privateKey)
		if err != nil {
			fmt.Println("failed to centalize: ", err)
			return
		}

		// save transaction
		err = SaveATransaction(userinfo["address"], *hash, "2", userinfo["pendingbalance"])
		if err != nil {
			fmt.Println("fail to save transaction: ", err)
			return
		}

		fmt.Printf("[%s] transaction submitted: %v\n", userinfo["address"], *hash)
	}
}


myEthereum/utils.go
package main

import (
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/keystore"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/ethclient"
	hdwallet "github.com/miguelmota/go-ethereum-hdwallet"
)

// MyTransaction : a transaction on the chain
// transaction type
//  0: recharge
//  1: withdraw
//  2: centralize
// transaction statuse
//  0: pending
//  1: done
type MyTransaction struct {
	Hash   string `json:"hash"`
	Type   string `json:"type"`
	Status string `json:"status"`
	Amount string `json:"amount"`
}

const (
	// AcountPoolPath : the file storing account pool information
	AcountPoolPath = `.\SystemData\addresses.txt`
	// AccountInfoPath : the file storing account information
	AccountInfoPath = `.\SystemData\AccountInfo\`
	// RPCAddress : the rpc address of the chain
	RPCAddress = "http://localhost:8545"
	// MainAddress : main address of the system
	MainAddress = "0xc0093215bec3cbb9522352dcb4e3fa8fd5b665d1"
)

// ReadFileContent : returns the file content as json
func ReadFileContent(path string) (*map[string]interface{}, error) {
	// open file
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}

	defer file.Close()

	// read file content
	buf := make([]byte, 1024)
	data := make([]byte, 0)

	for {
		n, err := file.Read(buf)
		if err != nil {
			if err == io.EOF {
				break
			} else {
				return nil, err
			}
		}

		data = append(data, buf[:n]...)
	}

	var datajson map[string]interface{}
	err = json.Unmarshal(data, &datajson)
	if err != nil {
		return nil, err
	}
	return &datajson, nil
}

// DecomposeTransaction : returns a MyTransaction
func DecomposeTransaction(client *ethclient.Client, tx *types.Transaction, tp string) (*MyTransaction, error) {
	hash := tx.Hash().Hex()
	amount := tx.Value().String()

	return &MyTransaction{Hash: strings.ToLower(hash), Amount: amount, Type: tp, Status: "0"}, nil
}

// MapToTransaction : transfer a map to a MyTransactions
func MapToTransaction(txmap map[string]interface{}) (*MyTransaction, error) {
	var tp, status, hash, amount string
	if value, ok := txmap["type"].(string); ok {
		tp = value
	}
	if value, ok := txmap["status"].(string); ok {
		status = value
	}
	if value, ok := txmap["hash"].(string); ok {
		hash = value
	}
	if value, ok := txmap["amount"].(string); ok {
		amount = value
	}
	if tp != "" && status != "" && hash != "" && amount != "" {
		return &MyTransaction{
			Hash:   hash,
			Type:   tp,
			Status: status,
			Amount: amount,
		}, nil
	}
	return nil, errors.New("wrong map format")
}

// TransactionToMap : transfer a MyTransactions to a map
func TransactionToMap(tx *MyTransaction) map[string]interface{} {
	txmap := make(map[string]interface{})
	txmap["hash"] = tx.Hash
	txmap["type"] = tx.Type
	txmap["status"] = tx.Status
	txmap["amount"] = tx.Amount
	return txmap
}

// MnemonicToAccount : returns the account address from a mnemonic.
func MnemonicToAccount(mnemonic string) (*string, error) {
	wallet, err := hdwallet.NewFromMnemonic(mnemonic)
	if err != nil {
		return nil, errors.New("failed to get account:" + err.Error())
	}

	path := hdwallet.MustParseDerivationPath("m/44'/60'/0'/0/0")
	account, err := wallet.Derive(path, false)
	if err != nil {
		fmt.Println("failed to generate wallet:", err)
		return nil, errors.New("failed to get account:" + err.Error())
	}

	address := account.Address.Hex()
	return &address, nil
}

// GetPrivateKey : get privatekey from keystore
func GetPrivateKey(privateKeyFile, password *string) (*string, error) {
	keyJSON, err := ioutil.ReadFile(*privateKeyFile)
	if err != nil {
		return nil, err
	}

	unlockedKey, err := keystore.DecryptKey(keyJSON, *password)
	if err != nil {
		return nil, err
	}

	privKey := hex.EncodeToString(unlockedKey.PrivateKey.D.Bytes())

	return &privKey, nil
}


myEthereum/system.go
package main

import (
	"context"
	"crypto/ecdsa"
	"encoding/json"
	"errors"
	"fmt"
	"math/big"
	"os"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/keystore"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/ethclient"
	hdwallet "github.com/miguelmota/go-ethereum-hdwallet"
	"github.com/tyler-smith/go-bip39"
)

// GenerateAccount : returns the mnemontic of the user
func GenerateAccount() (*string, error) {
	// generate a new mnemonic
	entropy, err := bip39.NewEntropy(256)
	if err != nil {
		return nil, errors.New("fail to generate mnemonic:" + err.Error())
	}

	mnemonic, err := bip39.NewMnemonic(entropy)
	if err != nil {
		return nil, errors.New("fail to generate mnemonic:" + err.Error())
	}

	// generate a wallet
	wallet, err := hdwallet.NewFromMnemonic(mnemonic)
	if err != nil {
		return nil, errors.New("fail to generate account:" + err.Error())
	}

	path := hdwallet.MustParseDerivationPath("m/44'/60'/0'/0/0")
	account, err := wallet.Derive(path, false)
	if err != nil {
		return nil, errors.New("fail to generate account:" + err.Error())
	}

	// store key
	ks := keystore.NewKeyStore("../data/keystore", keystore.StandardScryptN, keystore.StandardScryptP)

	privateKey, err := wallet.PrivateKey(account)
	if err != nil {
		return nil, errors.New("fail to generate account:" + err.Error())
	}

	_, err = ks.ImportECDSA(privateKey, "admin")
	if err != nil {
		return nil, errors.New("fail to generate account:" + err.Error())
	}

	return &mnemonic, nil
}

// RegisterAnAccount : returns the address and mnemonic of an available account
func RegisterAnAccount() (*string, *string, error) {
	// open pool file
	accountdataptr, err := ReadFileContent(AcountPoolPath)
	if err != nil {
		return nil, nil, errors.New("fail to open account pool file: " + err.Error())
	}

	accountdata := *accountdataptr

	accounts, ok := accountdata["accounts"].([]interface{})
	if !ok {
		return nil, nil, errors.New("fail to open account pool file")
	}

	// get an available address
	var tgaddress, tgmnemonic string
	newaccounts := []interface{}{}
	found := false
	for _, v := range accounts {
		account, ok := v.(map[string]interface{})
		if !ok {
			return nil, nil, errors.New("fail to open account pool file")
		}

		if account["status"] == "0" && !found {
			found = true
			account["status"] = "1"
			tgaddress, _ = account["address"].(string)
			tgmnemonic, _ = account["mnemonic"].(string)
		}

		newaccounts = append(newaccounts, account)
	}

	if !found {
		return nil, nil, errors.New("no available account")
	}

	accountdata["accounts"] = newaccounts

	newaccountdata, err := json.MarshalIndent(accountdata, "", "")
	if err != nil {
		return nil, nil, errors.New("fail to write account pool file")
	}

	// rewrite account info
	file, err := os.Create(AcountPoolPath)
	if err != nil {
		return nil, nil, errors.New("fail to write account pool file: " + err.Error())
	}

	defer file.Close()

	_, err = file.WriteString(string(newaccountdata))
	if err != nil {
		return nil, nil, errors.New("fail to write account pool file: " + err.Error())
	}

	return &tgaddress, &tgmnemonic, nil
}

// LoginAnAccount : log in the system
func LoginAnAccount(mnemonic string) bool {
	// get address
	address, err := MnemonicToAccount(mnemonic)
	if err != nil {
		return false
	}

	curuser = strings.ToLower(*address)
	return true
}

// RefreshAccountInfo : monitor transactions by hash and update current user's information file
// returns pending balance
func RefreshAccountInfo(client *ethclient.Client, address string) (*string, *string, error) {
	// read info file content
	dataptr, err := ReadFileContent(AccountInfoPath + address + ".txt")
	if err != nil {
		return nil, nil, errors.New("fail to open account info file: " + err.Error())
	}

	data := *dataptr

	// get balance, addrbalance and transactions
	var balance, addrbalance, pendingbalance *big.Int
	if value, ok := data["balance"].(string); ok {
		balance = new(big.Int)
		balance, _ = balance.SetString(value, 10)
	}
	if value, ok := data["addrbalance"].(string); ok {
		addrbalance = new(big.Int)
		addrbalance, _ = addrbalance.SetString(value, 10)
	}
	if value, ok := data["pendingbalance"].(string); ok {
		pendingbalance = new(big.Int)
		pendingbalance, _ = pendingbalance.SetString(value, 10)
	}
	var transactions []interface{}
	if value, ok := data["transactions"].([]interface{}); ok {
		transactions = value
	}

	// check all transaction by hash and refresh account info
	var newtransactions []interface{}
	haspending := false
	for _, t := range transactions {
		txmap, _ := t.(map[string]interface{})
		tx, err := MapToTransaction(txmap)
		if err != nil {
			return nil, nil, errors.New("fail to get transaction: " + err.Error())
		}

		if tx.Status == "0" {
			// get transaction by hash
			txHash := common.HexToHash(tx.Hash)
			_, ispending, err := client.TransactionByHash(context.Background(), txHash)
			if err != nil {
				return nil, nil, errors.New("fail to get transaction: " + err.Error())
			}

			// refresh transaction status
			if !ispending {
				tx.Status = "1"
				switch tx.Type {
				case "0":
					amount := new(big.Int)
					amount, ok := amount.SetString(tx.Amount, 10)
					if !ok {
						return nil, nil, errors.New("fail to get amount")
					}
					balance.Add(balance, amount)
					addrbalance.Add(addrbalance, amount)
				case "1":
					amount := new(big.Int)
					amount, ok := amount.SetString(tx.Amount, 10)
					if !ok {
						return nil, nil, errors.New("fail to get amount")
					}
					balance.Sub(balance, amount)
					addrbalance.Sub(addrbalance, amount)
				case "2":
					amount := new(big.Int)
					amount, ok := amount.SetString(tx.Amount, 10)
					if !ok {
						return nil, nil, errors.New("fail to get amount")
					}
					addrbalance.Sub(addrbalance, amount)
				}
			} else {
				haspending = true
				if tx.Type == "1" {
					amount := new(big.Int)
					amount, ok := amount.SetString(tx.Amount, 10)
					if !ok {
						return nil, nil, errors.New("fail to get amount")
					}
					pendingbalance = pendingbalance.Sub(balance, amount)
				}
			}
			newtransactions = append(newtransactions, TransactionToMap(tx))
		} else {
			newtransactions = append(newtransactions, txmap)
		}
	}

	// sychronize balance and pending balance
	if !haspending {
		pendingbalance = balance
	}

	// generate new info
	newinfo := make(map[string]interface{})
	newinfo["balance"] = balance.String()
	newinfo["addrbalance"] = addrbalance.String()
	newinfo["pendingbalance"] = pendingbalance.String()
	newinfo["transactions"] = newtransactions

	newinfodata, err := json.MarshalIndent(newinfo, "", "")
	if err != nil {
		return nil, nil, errors.New("fail to write user file")
	}

	// save new data
	file, err := os.Create(AccountInfoPath + address + ".txt")
	if err != nil {
		return nil, nil, errors.New("fail to write user file: " + err.Error())
	}

	defer file.Close()

	_, err = file.WriteString(string(newinfodata))
	if err != nil {
		return nil, nil, errors.New("fail to write userfile: " + err.Error())
	}

	adbl := addrbalance.String()
	pbl := pendingbalance.String()
	return &adbl, &pbl, nil
}

// RefreshAllAccount : monitor transactions by hash and update all users' information file
func RefreshAllAccount(client *ethclient.Client) (*[]map[string]string, bool) {
	// open pool file
	accountdataptr, err := ReadFileContent(AcountPoolPath)
	if err != nil {
		return nil, false
	}

	accountdata := *accountdataptr

	accounts, ok := accountdata["accounts"].([]interface{})
	if !ok {
		return nil, false
	}

	// calculate all the balance
	info := []map[string]string{}
	for _, v := range accounts {
		userinfo := make(map[string]string)
		account, _ := v.(map[string]interface{})

		if acctype, _ := account["status"].(string); acctype != "1" {
			continue
		}

		address, _ := account["address"].(string)

		addressbalance, _, err := RefreshAccountInfo(client, address)
		if err != nil {
			fmt.Println(err)
			return nil, false
		}
		userinfo["address"] = address
		userinfo["addressbalance"] = *addressbalance

		info = append(info, userinfo)
	}

	return &info, true
}

// StartATransaction : start a transaction, returns the hash
func StartATransaction(client *ethclient.Client, value *big.Int, from, to string, privateKey *ecdsa.PrivateKey) (*string, error) {
	// generate transaction
	gasLimit := uint64(80000)

	gasPrice, err := client.SuggestGasPrice(context.Background())
	if err != nil {
		return nil, err
	}

	delta := new(big.Int)
	delta, _ = delta.SetString("5000000000", 10)
	gasPrice = gasPrice.Add(gasPrice, delta)

	fromAddress := common.HexToAddress(from)
	toAddress := common.HexToAddress(to)

	nonce, err := client.PendingNonceAt(context.Background(), fromAddress)
	if err != nil {
		return nil, err
	}

	var data []byte
	tx := types.NewTransaction(nonce, toAddress, value, gasLimit, gasPrice, data)

	// get chain id
	chainID, err := client.NetworkID(context.Background())
	if err != nil {
		return nil, err
	}

	// sign the transaction
	signedTx, err := types.SignTx(tx, types.NewEIP155Signer(chainID), privateKey)
	if err != nil {
		return nil, err
	}

	// send transactions
	err = client.SendTransaction(context.Background(), signedTx)
	if err != nil {
		return nil, err
	}

	hash := signedTx.Hash().Hex()

	return &hash, nil
}

// SaveATransaction : save a transaction to account info file
func SaveATransaction(address, hash, tp, amount string) error {
	// read account info
	dataptr, err := ReadFileContent(AccountInfoPath + address + ".txt")
	if err != nil {
		return errors.New(("fail to open account info file: " + err.Error()))
	}

	data := *dataptr

	// get balance, addrbalance and transactions
	var balance, addrbalance, pendingbalance *big.Int
	if value, ok := data["balance"].(string); ok {
		balance = new(big.Int)
		balance, _ = balance.SetString(value, 10)
	}
	if value, ok := data["addrbalance"].(string); ok {
		addrbalance = new(big.Int)
		addrbalance, _ = addrbalance.SetString(value, 10)
	}
	if value, ok := data["pendingbalance"].(string); ok {
		pendingbalance = new(big.Int)
		pendingbalance, _ = pendingbalance.SetString(value, 10)
	}
	var transactions []interface{}
	if value, ok := data["transactions"].([]interface{}); ok {
		transactions = value
	}

	transaction := MyTransaction{Hash: hash, Type: tp, Status: "0", Amount: amount}
	transactions = append(transactions, TransactionToMap(&transaction))

	// generate new info
	newinfo := make(map[string]interface{})
	newinfo["balance"] = balance.String()
	newinfo["addrbalance"] = addrbalance.String()
	newinfo["pendingbalance"] = pendingbalance.String()
	newinfo["transactions"] = transactions

	newinfodata, err := json.MarshalIndent(newinfo, "", "")
	if err != nil {
		return errors.New(("fail to write user file: " + err.Error()))
	}

	// save new data
	file, err := os.Create(AccountInfoPath + address + ".txt")
	if err != nil {
		return errors.New(("fail to write user file: " + err.Error()))
	}

	defer file.Close()

	_, err = file.WriteString(string(newinfodata))
	if err != nil {
		return errors.New(("fail to write user file: " + err.Error()))
	}

	return nil
}
